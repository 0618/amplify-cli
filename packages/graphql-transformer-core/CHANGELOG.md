# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.0.9"></a>
## [1.0.9](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.8...graphql-transformer-core@1.0.9) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.8"></a>
## [1.0.8](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.7...graphql-transformer-core@1.0.8) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.7"></a>
## [1.0.7](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.6...graphql-transformer-core@1.0.7) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.6"></a>
## [1.0.6](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.5...graphql-transformer-core@1.0.6) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.5"></a>
## [1.0.5](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.4...graphql-transformer-core@1.0.5) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.4"></a>
## 1.0.4 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.3"></a>
## 1.0.3 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.2"></a>
## 1.0.2 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.1"></a>
## 1.0.1 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core
