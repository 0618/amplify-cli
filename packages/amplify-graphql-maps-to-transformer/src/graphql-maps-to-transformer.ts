import { TransformerPluginBase, InvalidDirectiveError } from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerPluginType,
  TransformerSchemaVisitStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { ObjectTypeDefinitionNode, DirectiveNode } from 'graphql';
import { attachInputMappingSlot, attachResponseMappingSlot } from './field-mapping-resolvers';

const directiveName = 'mapsTo';

const directiveDefinition = `
  directive @${directiveName}(name: String!) on OBJECT
`;

export class MapsToTransformer extends TransformerPluginBase {
  constructor() {
    super(`amplify-maps-to-transformer`, directiveDefinition, TransformerPluginType.GENERIC);
  }

  object = (definition: ObjectTypeDefinitionNode, directive: DirectiveNode, ctx: TransformerSchemaVisitStepContextProvider) => {
    const modelName = definition.name.value;
    const prevNameNode = directive.arguments?.find(arg => arg.name.value === 'name');
    if (!prevNameNode) {
      throw new InvalidDirectiveError(`name is required in @${directiveName} directive`);
    }

    if (prevNameNode.value.kind !== 'StringValue') {
      throw new InvalidDirectiveError(`A single string must be provided for "name" in @${directiveName} directive`);
    }

    const originalName = prevNameNode.value.value;
    ctx.resourceHelper.setModelNameMapping(modelName, originalName);
  };

  generateResolvers = (context: TransformerContextProvider) =>
    context.resourceHelper.getResolverMapRegistry().forEach(({ resolverTypeName, resolverFieldName, fieldMap, isResultList }) => {
      const resolver = context.resolvers.getResolver(resolverTypeName, resolverFieldName);
      if (!resolver) {
        return;
      }
      if (resolverTypeName === 'Mutation') {
        attachInputMappingSlot({ resolver, resolverFieldName, resolverTypeName, fieldMap });
        attachResponseMappingSlot({ slotName: 'postUpdate', resolver, resolverFieldName, resolverTypeName, fieldMap, isList: false });
        return;
      }
      attachResponseMappingSlot({
        slotName: 'postDataLoad',
        resolver,
        resolverFieldName,
        resolverTypeName,
        fieldMap: fieldMap,
        isList: isResultList,
      });
    });
}
