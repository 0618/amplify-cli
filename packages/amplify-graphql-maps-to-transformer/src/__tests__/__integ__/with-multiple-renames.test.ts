import { ConflictHandlerType, GraphQLTransform } from '@aws-amplify/graphql-transformer-core';
import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { BelongsToTransformer, HasManyTransformer, HasOneTransformer } from '@aws-amplify/graphql-relational-transformer';
import { MapsToTransformer } from '../../graphql-maps-to-transformer';

const originalSchema = /* GraphQL */ `
  type Checklist @model {
    id: ID!
    tasks: [Task] @hasMany
  }

  type Task @model {
    id: ID!
    title: String!
    loc: Location @hasOne
    day: Day @belongsTo
  }

  type Location @model {
    id: ID!
    name: String
  }

  type Day @model {
    id: ID!
    task: Task @hasOne
  }
`;

const multipleForeignKeyRenames = /* GraphQL */ `
  type Agenda @model @mapsTo(name: "Checklist") {
    id: ID!
    tasks: [Todo] @hasMany
  }

  type Todo @model @mapsTo(name: "Task") {
    id: ID!
    title: String!
    loc: Location @hasOne
    day: Day @belongsTo
  }

  type Location @model {
    id: ID!
    name: String
  }

  type Day @model {
    id: ID!
    task: Todo @hasOne
  }
`;

const transformSchema = (schema: string, enableDataStore = false) => {
  const transformer = new GraphQLTransform({
    transformers: [
      new ModelTransformer(),
      new HasManyTransformer(),
      new HasOneTransformer(),
      new BelongsToTransformer(),
      new MapsToTransformer(),
    ],
    sandboxModeEnabled: true,
    resolverConfig: enableDataStore
      ? {
          project: {
            ConflictDetection: 'VERSION',
            ConflictHandler: ConflictHandlerType.AUTOMERGE,
          },
        }
      : undefined,
  });
  return transformer.transform(schema);
};

const crudResolvers = (modelName: string) => [
  `Mutation.create${modelName}.postUpdate.1.res.vtl`,
  `Mutation.create${modelName}.init.2.req.vtl`,
  `Mutation.update${modelName}.postUpdate.1.res.vtl`,
  `Mutation.update${modelName}.init.2.req.vtl`,
  `Query.get${modelName}.postDataLoad.1.res.vtl`,
  `Query.list${modelName}s.postDataLoad.1.res.vtl`,
];

test('@mapsTo with multiple foreign key field renames on single model', () => {
  const out = transformSchema(multipleForeignKeyRenames);
  const expectedUndefinedResovlers = ['Location', 'Day', 'Agenda'].flatMap(crudResolvers);
  expectedUndefinedResovlers.forEach(resolver => expect(out.resolvers[resolver]).toBeUndefined());

  const expectedRemappingResolvers = ['Todo'].flatMap(crudResolvers);
  expectedRemappingResolvers.forEach(resolver => {
    expect(out.resolvers[resolver]).toMatchSnapshot();
  });
});

test('maps-to-transformer does not apply any resolvers to non-mapped models', () => {
  const out = transformSchema(originalSchema);
  const expectUndefinedResolversList = ['Checklist', 'Task', 'Location', 'Day'].flatMap(crudResolvers);
  expectUndefinedResolversList.forEach(resolver => expect(out.resolvers[resolver]).toBeUndefined());
});

test('maps sync resolvers when DataStore is enabled', () => {
  const out = transformSchema(multipleForeignKeyRenames, true);
  const expectedSyncResolvers = ['Todo'].map(type => `Query.sync${type}s.postDataLoad.1.res.vtl`);
  expectedSyncResolvers.forEach(resolver => expect(out.resolvers[resolver]).toMatchSnapshot());
});
