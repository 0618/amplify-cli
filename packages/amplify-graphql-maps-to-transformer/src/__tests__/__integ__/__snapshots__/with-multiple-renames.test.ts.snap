// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@mapsTo with multiple foreign key field renames on single model 1`] = `
"$util.qr($ctx.prev.result.put(\\"agendaTasksId\\", $ctx.prev.result.checklistTasksId))
$util.qr($ctx.prev.result.remove(\\"checklistTasksId\\"))
$util.qr($ctx.prev.result.put(\\"todoLocId\\", $ctx.prev.result.taskLocId))
$util.qr($ctx.prev.result.remove(\\"taskLocId\\"))
$util.qr($ctx.prev.result.put(\\"todoDayId\\", $ctx.prev.result.taskDayId))
$util.qr($ctx.prev.result.remove(\\"taskDayId\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`@mapsTo with multiple foreign key field renames on single model 2`] = `
"$util.qr($ctx.args.input.put(\\"checklistTasksId\\", $ctx.args.input.agendaTasksId))
$util.qr($ctx.args.input.remove(\\"agendaTasksId\\"))
$util.qr($ctx.args.input.put(\\"taskLocId\\", $ctx.args.input.todoLocId))
$util.qr($ctx.args.input.remove(\\"todoLocId\\"))
$util.qr($ctx.args.input.put(\\"taskDayId\\", $ctx.args.input.todoDayId))
$util.qr($ctx.args.input.remove(\\"todoDayId\\"))
$util.toJson({})"
`;

exports[`@mapsTo with multiple foreign key field renames on single model 3`] = `
"$util.qr($ctx.prev.result.put(\\"agendaTasksId\\", $ctx.prev.result.checklistTasksId))
$util.qr($ctx.prev.result.remove(\\"checklistTasksId\\"))
$util.qr($ctx.prev.result.put(\\"todoLocId\\", $ctx.prev.result.taskLocId))
$util.qr($ctx.prev.result.remove(\\"taskLocId\\"))
$util.qr($ctx.prev.result.put(\\"todoDayId\\", $ctx.prev.result.taskDayId))
$util.qr($ctx.prev.result.remove(\\"taskDayId\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`@mapsTo with multiple foreign key field renames on single model 4`] = `
"$util.qr($ctx.args.input.put(\\"checklistTasksId\\", $ctx.args.input.agendaTasksId))
$util.qr($ctx.args.input.remove(\\"agendaTasksId\\"))
$util.qr($ctx.args.input.put(\\"taskLocId\\", $ctx.args.input.todoLocId))
$util.qr($ctx.args.input.remove(\\"todoLocId\\"))
$util.qr($ctx.args.input.put(\\"taskDayId\\", $ctx.args.input.todoDayId))
$util.qr($ctx.args.input.remove(\\"todoDayId\\"))
$util.toJson({})"
`;

exports[`@mapsTo with multiple foreign key field renames on single model 5`] = `
"$util.qr($ctx.prev.result.put(\\"agendaTasksId\\", $ctx.prev.result.checklistTasksId))
$util.qr($ctx.prev.result.remove(\\"checklistTasksId\\"))
$util.qr($ctx.prev.result.put(\\"todoLocId\\", $ctx.prev.result.taskLocId))
$util.qr($ctx.prev.result.remove(\\"taskLocId\\"))
$util.qr($ctx.prev.result.put(\\"todoDayId\\", $ctx.prev.result.taskDayId))
$util.qr($ctx.prev.result.remove(\\"taskDayId\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`@mapsTo with multiple foreign key field renames on single model 6`] = `
"#foreach( $item in $ctx.prev.result.items )
  $util.qr($item.put(\\"agendaTasksId\\", $item.checklistTasksId))
  $util.qr($item.remove(\\"checklistTasksId\\"))
  $util.qr($item.put(\\"todoLocId\\", $item.taskLocId))
  $util.qr($item.remove(\\"taskLocId\\"))
  $util.qr($item.put(\\"todoDayId\\", $item.taskDayId))
  $util.qr($item.remove(\\"taskDayId\\"))
#end
$util.toJson($ctx.prev.result)"
`;

exports[`maps sync resolvers when DataStore is enabled 1`] = `
"#foreach( $item in $ctx.prev.result.items )
  $util.qr($item.put(\\"agendaTasksId\\", $item.checklistTasksId))
  $util.qr($item.remove(\\"checklistTasksId\\"))
  $util.qr($item.put(\\"todoLocId\\", $item.taskLocId))
  $util.qr($item.remove(\\"taskLocId\\"))
  $util.qr($item.put(\\"todoDayId\\", $item.taskDayId))
  $util.qr($item.remove(\\"taskDayId\\"))
#end
$util.toJson($ctx.prev.result)"
`;
